File: src/main/java/cn/wjybxx/dson/types/OffsetTimestamp.java
package cn.wjybxx.dson.types;

import cn.wjybxx.dson.DsonLites;
import cn.wjybxx.dson.internal.InternalUtils;

import javax.annotation.concurrent.Immutable;
import java.time.LocalDate;
import java.time.LocalDateTime;
import java.time.LocalTime;
import java.time.ZoneOffset;
import java.time.format.DateTimeFormatter;

/**
 * 有时区偏移的时间戳
 *
 * @author wjybxx
 * date - 2023/6/17
 */
@Immutable
public final class OffsetTimestamp {

    public static final int MASK_DATE = 1;
    public static final int MASK_TIME = 1 << 1;
    public static final int MASK_OFFSET = 1 << 2;
    public static final int MASK_NANOS = 1 << 3;

    public static final int MASK_DATETIME = MASK_DATE | MASK_TIME;
    public static final int MASK_OFFSET_DATETIME = MASK_DATE | MASK_TIME | MASK_OFFSET;

    private final long seconds;
    private final int nanos;
    private final int offset;
    private final int enables;

    public OffsetTimestamp(long seconds) {
        this(seconds, 0, 0, MASK_DATETIME);
    }

    /**
     * @param seconds 纪元时间-秒
     * @param nanos   纪元时间的纳秒部分
     * @param offset  时区偏移-秒
     * @param enables 哪些字段有效
     */
    public OffsetTimestamp(long seconds, int nanos, int offset, int enables) {
        if (InternalUtils.isDisabled(enables, MASK_DATE) && InternalUtils.isDisabled(enables, MASK_TIME)) {
            throw new IllegalArgumentException("date and time are disabled");
        }
        if (offset != 0 && InternalUtils.isDisabled(enables, MASK_OFFSET)) {
            throw new IllegalArgumentException("offset is disabled, but the value is not 0");
        }
        if (nanos != 0 && InternalUtils.isDisabled(enables, MASK_NANOS)) {
            throw new IllegalArgumentException("nanos is disabled, but the value is not 0");
        }
        if (nanos > 999_999_999 || nanos < 0) {
            throw new IllegalArgumentException("nanos > 999999999 or < 0");
        }
        this.seconds = seconds;
        this.nanos = nanos;
        this.offset = offset;
        this.enables = enables;
    }

    public long getSeconds() {
        return seconds;
    }

    public int getNanos() {
        return nanos;
    }

    public int getOffset() {
        return offset;
    }

    public int getEnables() {
        return enables;
    }

    // region

    public boolean hasDate() {
        return InternalUtils.isEnabled(enables, MASK_DATE);
    }

    public boolean hasTime() {
        return InternalUtils.isEnabled(enables, MASK_TIME);
    }

    public boolean hasOffset() {
        return InternalUtils.isEnabled(enables, MASK_OFFSET);
    }

    public boolean hasNanos() {
        return InternalUtils.isEnabled(enables, MASK_NANOS);
    }

    public boolean hasFields(int mask) {
        return InternalUtils.isEnabled(enables, mask);
    }

    public boolean canConvertNanosToMillis() {
        return (nanos % 1000_000) == 0;
    }

    public int getMillisOfNanos() {
        return nanos / 1000_000;
    }

    /** @return 固定格式 yyyy-MM-dd */
    public static String formatDate(long seconds) {
        return LocalDateTime.ofEpochSecond(seconds, 0, ZoneOffset.UTC)
                .toLocalDate()
                .toString();
    }

    /** @return 固定格式 HH:mm:ss */
    public static String formatTime(long seconds) {
        return LocalDateTime.ofEpochSecond(seconds, 1, ZoneOffset.UTC)
                .toLocalTime()
                .toString()
                .substring(0, 8);
    }

    /** @return 固定格式 yyyy-MM-dd'T'HH:mm:ss */
    public static String formatDateTime(long seconds) {
        return formatDate(seconds) + "T" + formatTime(seconds);
    }

    /**
     * Z
     * +HH:mm
     * +HH:mm:ss
     */
    public static String formatOffset(int offset) {
        if (offset == 0) {
            return "Z";
        }
        String pre = offset < 0 ? "-" : "+";
        return pre + LocalTime.ofSecondOfDay(offset).toString();
    }

    /** @param dateString 限定格式 yyyy-MM-dd */
    public static LocalDate parseDate(String dateString) {
//        if (dateString.length() != 10) throw new IllegalArgumentException("invalid dateString " + dateString);
        return LocalDate.parse(dateString, DateTimeFormatter.ISO_DATE);
    }

    /** @param timeString 限定格式 HH:mm:ss */
    public static LocalTime parseTime(String timeString) {
        if (timeString.length() != 8) throw new IllegalArgumentException("invalid timeString " + timeString);
        return LocalTime.parse(timeString, DateTimeFormatter.ISO_TIME);
    }

    /** @param timeString 限定格式 yyyy-MM-dd'T'HH:mm:ss */
    public static LocalDateTime parseDateTime(String timeString) {
        return LocalDateTime.parse(timeString, DateTimeFormatter.ISO_DATE_TIME);
    }

    /**
     * Z
     * +H
     * +HH
     * +HH:mm
     * +HH:mm:ss
     */
    public static int parseOffset(String offsetString) {
        return switch (offsetString.length()) {
            case 1, 2, 3, 6, 9 -> ZoneOffset.of(offsetString).getTotalSeconds();
            default -> throw new IllegalArgumentException("invalid offset string " + offsetString);
        };
    }

    // endregion

    //region equals

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (o == null || getClass() != o.getClass()) return false;

        OffsetTimestamp offsetTimestamp = (OffsetTimestamp) o;

        if (seconds != offsetTimestamp.seconds) return false;
        if (nanos != offsetTimestamp.nanos) return false;
        if (offset != offsetTimestamp.offset) return false;
        return enables == offsetTimestamp.enables;
    }

    @Override
    public int hashCode() {
        int result = (int) (seconds ^ (seconds >>> 32));
        result = 31 * result + nanos;
        result = 31 * result + offset;
        result = 31 * result + enables;
        return result;
    }
    // endregion

    @Override
    public String toString() {
        StringBuilder sb = new StringBuilder();
        sb.append("OffsetTimestamp{");
        if (hasDate()) {
            sb.append("date: '").append(formatDate(seconds));
        }
        if (hasTime()) {
            if (hasDate()) {
                sb.append(", ");
            }
            sb.append("time: '").append(formatTime(seconds));
        }
        if (nanos != 0) {
            sb.append(", ");
            if (canConvertNanosToMillis()) {
                sb.append("millis: ").append(getMillisOfNanos());
            } else {
                sb.append("nanos: ").append(nanos);
            }
        }
        if (hasOffset()) {
            sb.append(", ");
            sb.append("offset: '").append(formatOffset(offset))
                    .append("'");
        }
        return sb.append('}')
                .toString();
    }

    //
    public static final String NAMES_DATE = "date";
    public static final String NAMES_TIME = "time";
    public static final String NAMES_MILLIS = "millis";

    public static final String NAMES_SECONDS = "seconds";
    public static final String NAMES_NANOS = "nanos";
    public static final String NAMES_OFFSET = "offset";
    public static final String NAMES_ENABLES = "enables";

    public static final int NUMBERS_SECONDS = DsonLites.makeFullNumberZeroIdep(0);
    public static final int NUMBERS_NANOS = DsonLites.makeFullNumberZeroIdep(1);
    public static final int NUMBERS_OFFSET = DsonLites.makeFullNumberZeroIdep(2);
    public static final int NUMBERS_ENABLES = DsonLites.makeFullNumberZeroIdep(3);
}

File: csharp/Dson/Types/OffsetTimestamp.cs
/*
 * Copyright 2023 wjybxx(845740757@qq.com)
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

using System.Globalization;
using System.Text;
using Dson.Text;

namespace Dson;

/// <summary>
/// 带时区偏移的时间戳
/// </summary>
public struct OffsetTimestamp : IEquatable<OffsetTimestamp>
{
    public const int MASK_DATE = 1;
    public const int MASK_TIME = 1 << 1;
    public const int MASK_OFFSET = 1 << 2;
    public const int MASK_NANOS = 1 << 3;

    public const int MASK_DATETIME = MASK_DATE | MASK_TIME;
    public const int MASK_OFFSET_DATETIME = MASK_DATE | MASK_TIME | MASK_OFFSET;

    public readonly long Seconds;
    public readonly int Nanos;
    public readonly int Offset;
    public readonly int Enables;

    public OffsetTimestamp(long seconds)
        : this(seconds, 0, 0, MASK_DATETIME) {
    }

    /// <summary>
    /// 
    /// </summary>
    /// <param name="seconds">纪元时间秒时间戳</param>
    /// <param name="nanos">时间戳的纳秒部分</param>
    /// <param name="offset">时区偏移量--秒</param>
    /// <param name="enables">启用的字段信息</param>
    public OffsetTimestamp(long seconds, int nanos, int offset, int enables) {
        this.Seconds = seconds;
        this.Nanos = nanos;
        this.Offset = offset;
        this.Enables = enables;
    }

    public bool hasDate() {
        return DsonInternals.isEnabled(Enables, MASK_DATE);
    }

    public bool hasTime() {
        return DsonInternals.isEnabled(Enables, MASK_TIME);
    }

    public bool hasOffset() {
        return DsonInternals.isEnabled(Enables, MASK_OFFSET);
    }

    public bool canConvertNanosToMillis() {
        return (Nanos % 1000_000) == 0;
    }

    public int convertNanosToMillis() {
        return Nanos / 1000_000;
    }

    #region equals

    public bool Equals(OffsetTimestamp other) {
        return Seconds == other.Seconds && Nanos == other.Nanos && Offset == other.Offset && Enables == other.Enables;
    }

    public override bool Equals(object? obj) {
        return obj is OffsetTimestamp other && Equals(other);
    }

    public override int GetHashCode() {
        return HashCode.Combine(Seconds, Nanos, Offset, Enables);
    }

    public static bool operator ==(OffsetTimestamp left, OffsetTimestamp right) {
        return left.Equals(right);
    }

    public static bool operator !=(OffsetTimestamp left, OffsetTimestamp right) {
        return !left.Equals(right);
    }

    #endregion

    public override string ToString() {
        return $"{nameof(Seconds)}: {Seconds}, {nameof(Nanos)}: {Nanos}, {nameof(Offset)}: {Offset}, {nameof(Enables)}: {Enables}";
    }

    #region 解析

    public static DateTime parseDateTime(string datetimeString) {
        return DateTime.ParseExact(datetimeString, "yyyy-MM-ddTHH:mm:ss", CultureInfo.InvariantCulture);
    }

    public static DateOnly parseDate(string dateString) {
        return DateOnly.ParseExact(dateString, "yyyy-MM-dd", CultureInfo.InvariantCulture);
    }

    public static TimeOnly parseTime(string timeString) {
        return TimeOnly.ParseExact(timeString, "HH:mm:ss", CultureInfo.InvariantCulture);
    }

    public static int parseOffset(string offsetString) {
        if (offsetString == "Z" || offsetString == "z") {
            return 0;
        }
        if (offsetString[0] != '+' && offsetString[0] != '-') {
            throw new DsonParseException("Invalid offsetString, plus/minus not found when expected: " + offsetString);
        }
        // 不想写得太复杂，补全后解析
        switch (offsetString.Length) {
            case 2: { // ±H
                offsetString = offsetString + "0:00:00";
                break;
            }
            case 3: { // ±HH
                offsetString = offsetString + ":00:00";
                break;
            }
            case 5: { // ±H:mm
                offsetString = new StringBuilder(offsetString, 9)
                    .Insert(1, '0')
                    .Append(":00")
                    .ToString();
                break;
            }
            case 6: { // ±HH:mm
                offsetString = offsetString + ":00";
                break;
            }
            case 9: { // ±HH:mm:ss
                break;
            }
        }
        long seconds = (parseTime(offsetString.Substring(1)).Ticks / DsonInternals.TicksPerSecond);
        if (offsetString[0] == '+') {
            return (int)seconds;
        }
        return (int)(-1 * seconds);
    }

    public static string formatDateTime(long seconds) {
        return DateTime.UnixEpoch.AddSeconds(seconds).ToString("s");
    }

    public static string formatDate(long epochSeconds) {
        string fullDate = DateTime.UnixEpoch.AddSeconds(epochSeconds).ToString("s");
        return fullDate.Substring(0, fullDate.IndexOf('T'));
    }

    public static string formatTime(long epochSeconds) {
        string fullDate = DateTime.UnixEpoch.AddSeconds(epochSeconds).ToString("s");
        return fullDate.Substring(fullDate.IndexOf('T') + 1);
    }

    public static string formatOffset(int offsetSeconds) {
        if (offsetSeconds == 0) {
            return "Z";
        }
        string sign = offsetSeconds < 0 ? "-" : "+";
        if (offsetSeconds % 60 == 0) { // 没有秒部分
            return sign + formatTime(Math.Abs(offsetSeconds)).Substring(0, 5);
        }
        else {
            return sign + formatTime(Math.Abs(offsetSeconds));
        }
    }

    #endregion


    #region 常量

    public const string NAMES_DATE = "date";
    public const string NAMES_TIME = "time";
    public const string NAMES_MILLIS = "millis";

    public const string NAMES_SECONDS = "seconds";
    public const string NAMES_NANOS = "nanos";
    public const string NAMES_OFFSET = "offset";
    public const string NAMES_ENABLES = "enables";

    public static readonly int NUMBERS_SECONDS = Dsons.MakeFullNumberZeroIdep(0);
    public static readonly int NUMBERS_NANOS = Dsons.MakeFullNumberZeroIdep(1);
    public static readonly int NUMBERS_OFFSET = Dsons.MakeFullNumberZeroIdep(2);
    public static readonly int NUMBERS_ENABLES = Dsons.MakeFullNumberZeroIdep(3);

    #endregion
}